### PR TITLE
free_layer_convergence prepare and analyse

### PR DESCRIPTION
**Pull request by KaworuIkari**

### CODE REVIEW

**Comment by KaworuIkari** [comment]
there is a bug in calculating the displacement. I think before solving it you can check whether the method I used to calculate displacement is ok or not

**Comment by Matt-Ord** [comment]
By displacement which function do you mean exactly - building or analothe slab?

File: pseudopotential_calculator/calculations/slab.py [hunk 1]
```diff
@@ -82,17 +83,23 @@ def get_surface(
...
-    slab = cast(Atoms, surface(atom, slab_direction, n_layer, 0))
+    n_layer = n_free_layer + n_fixed_layer
+    atoms = cast(Atoms, surface(atom, slab_direction, n_layer, 0))
 
     height_per_layer = _get_height_per_layer(atom, slab_direction)
-    add_vacuum(slab, n_vacuum_layer * height_per_layer)
-    return slab
+    add_vacuum(atoms, n_vacuum_layer * height_per_layer)
+
+    # TODO: fix the first n_fixed atoms, and fix the x,y coordinate of the
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:96]: Left this todo for you

File: pseudopotential_calculator/calculations/slab.py [hunk 2]
```diff
@@ -119,11 +126,56 @@ def plot_energy_against_n_vacuum_layer(
...
         ax=ax,
     )
+
+
+def get_atom_displacement(atom: Atoms, n_th_layer: int) -> float:
+    heights = cast(list[float], atom.positions[:, 2])  # type: ignore bad lib
+    n_cu_layer = len(np.unique(heights))
+    sorted_heights = np.unique(heights)[::-1]
+    atom_height = sorted_heights[n_th_layer - 1]
+    height_per_fixed_layer = sorted_heights[-2]
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:144]: split this height logic into another function and share with the vacuum convergence plot

File: pseudopotential_calculator/calculations/slab.py [hunk 3]
```diff
@@ -119,11 +126,56 @@ def plot_energy_against_n_vacuum_layer(
...
 
     return plot_data_comparison(
         ("Vacuum Height", np.array(n_vacuum_layer), "N layers"),
         ("Energy", np.array(energies), "J"),
         ax=ax,
     )
+
+
+def get_atom_displacement(atom: Atoms, n_th_layer: int) -> float:
+    heights = cast(list[float], atom.positions[:, 2])  # type: ignore bad lib
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:140]: this is not a list?

File: scripts/copper/slab/analyze.py [hunk 4]
```diff
@@ -25,5 +27,24 @@ def _analyze_convergence_with_n_vacuum_layer() -> None:
     save_fig(fig, SAVE_DIR / plot_name)
 
 
+def _analyze_convergence_with_n_free_layer() -> None:
+    data_path = FREE_LAYER_PATH
+    maybe_copy_files_from_hpc(data_path, data_path)
+    calculators = load_all_calculators(data_path)
+    fig, ax = get_figure(None)
+    for n_th_layer in range(1, 6):
```
Comment (Matt-Ord) [inline scripts/copper/slab/analyze.py:35]: for layer in range():

File: pseudopotential_calculator/calculations/slab.py [hunk 5]
```diff
@@ -119,11 +126,56 @@ def plot_energy_against_n_vacuum_layer(
...
+        )
+        n_free_layers = get_n_free_layer(atom)
+        n_free_layer.append(n_free_layers)
+
+    fig, ax, line = plot_data_comparison(
+        ("N Free Layers", np.array(n_free_layer), None),
+        ("Displacement", np.array(displacement), "m"),
+        ax=ax,
+    )
+    line.set_label(f"layer {layer_idx}")
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:180]: my bad, maybe remove set_label here

File: scripts/copper/slab/analyze.py [hunk 6]
```diff
@@ -25,5 +27,24 @@ def _analyze_convergence_with_n_vacuum_layer() -> None:
...
+        _, _, line = plot_displacement_against_n_free_layer(
+            calculators,
+            n_th_layer,
+            ax=ax,
+        )
+        line.set_label(f"{n_th_layer}_th_layer")
```
Comment (Matt-Ord) [inline scripts/copper/slab/analyze.py:41]: just f"layer {n}"

**Review by Matt-Ord — CHANGES_REQUESTED** [review]
Really nice code! I left a few comments (I made a couple of tweaks myself but I left the hard changes for you to do 😄 )

File: pseudopotential_calculator/calculations/slab.py (hunk 3 above)
Comment (KaworuIkari) [reply pseudopotential_calculator/calculations/slab.py:140]: This is what the error reports
```
Type of "positions" is partially unknown
Type of "positions" is "Any | ndarray[Any, dtype[bool_]] | Unknown"
```

Comment (Matt-Ord) [reply pseudopotential_calculator/calculations/slab.py:140]: Yeah so cast it to an ndarray of np.dtype[np.float_]

File: pseudopotential_calculator/calculations/slab.py [hunk 7]
```diff
@@ -78,32 +79,74 @@ def _get_height_per_layer(
...
-    height_per_layer = slab_height / len(atom_heights)
+    slab_height = get_n_th_slab_layer_height(atom, 1)
+    height_per_layer = get_height_per_slab_layer(atom)
 
     vacuum_height = cell_height - slab_height
 
     return vacuum_height / height_per_layer
 
 
+def _get_heights_sorted(atom: Atoms) -> np.ndarray[Any, np.dtype[np.float64]]:
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:128]: get_sorted_heights is a better name

File: pseudopotential_calculator/calculations/slab.py [hunk 8]
```diff
@@ -78,32 +79,74 @@ def _get_height_per_layer(
...
+    heights = cast(np.ndarray[Any, np.dtype[np.float64]], atom.positions[:, 2])  # type: ignore bad lib
+    return np.unique(heights)[::-1]
+
+
+def get_n_th_slab_layer_height(atom: Atoms, n_th_layer: int) -> float:
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:133]: get_height_of_layer(atoms: Atoms, layer_idx: int) -> float:

Could you also replace all atom: Atoms with atoms: Atoms ...

File: pseudopotential_calculator/calculations/slab.py [hunk 9]
```diff
@@ -78,32 +79,74 @@ def _get_height_per_layer(
...
+    heights = _get_heights_sorted(atom)
+    return heights[n_th_layer - 1]
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:135]: heights[layer_idx]

File: pseudopotential_calculator/calculations/slab.py [hunk 10]
```diff
@@ -78,32 +79,74 @@ def _get_height_per_layer(
...
+
+
+def get_height_per_slab_layer(atom: Atoms) -> float:
+    _get_heights_sorted(atom)
+    heights = _get_heights_sorted(atom)
+    sorted_heights = np.unique(heights)[::-1]
+    return sorted_heights[-2]
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:142]: need to find the difference between heights here

File: pseudopotential_calculator/calculations/slab.py [hunk 11]
```diff
@@ -78,32 +79,74 @@ def _get_height_per_layer(
...
+    heights = cast(np.ndarray[Any, np.dtype[np.float64]], atom.positions[:, 2])  # type: ignore bad lib
+    return np.unique(heights)[::-1]
+
+
+def get_n_th_slab_layer_height(atom: Atoms, n_th_layer: int) -> float:
+    heights = _get_heights_sorted(atom)
+    return heights[n_th_layer - 1]
+
+
+def get_height_per_slab_layer(atom: Atoms) -> float:
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:138]: _get_height_par_layer(...)  this should be private

File: pseudopotential_calculator/calculations/slab.py [hunk 12]
```diff
@@ -78,32 +79,74 @@ def _get_height_per_layer(
...
+    _get_heights_sorted(atom)
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:139]: dead code

File: pseudopotential_calculator/calculations/slab.py [hunk 13]
```diff
@@ -78,32 +79,74 @@ def _get_height_per_layer(
...
+    heights = _get_heights_sorted(atom)
+    sorted_heights = np.unique(heights)[::-1]
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:141]: already sorted

File: pseudopotential_calculator/calculations/slab.py [hunk 14]
```diff
@@ -78,32 +79,74 @@ def _get_height_per_layer(
...
+    return sorted_heights[-2]
+
+
+def get_n_slab_layer(atom: Atoms) -> int:
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:145]: get_n_slab_layers

File: pseudopotential_calculator/calculations/slab.py [hunk 15]
```diff
@@ -78,32 +79,74 @@ def _get_height_per_layer(
...
+    heights = _get_heights_sorted(atom)
+    return len(heights)
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:147]: no need to assign to a variable first

File: pseudopotential_calculator/calculations/slab.py [hunk 16]
```diff
@@ -119,11 +162,45 @@ def plot_energy_against_n_vacuum_layer(
...
         n_vacuum_layer.append(n_vacuum_layers)
 
     return plot_data_comparison(
         ("Vacuum Height", np.array(n_vacuum_layer), "N layers"),
         ("Energy", np.array(energies), "J"),
         ax=ax,
     )
+
+
+def get_atom_displacement(atom: Atoms, n_th_layer: int) -> float:
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:175]: get_displacement_of_layer(... , layer_idx)

File: pseudopotential_calculator/calculations/slab.py [hunk 17]
```diff
@@ -119,11 +162,45 @@ def plot_energy_against_n_vacuum_layer(
...
+    n_slab_layer = get_n_slab_layer(atom)
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:176]: n_slab_layers =

File: pseudopotential_calculator/calculations/slab.py [hunk 18]
```diff
@@ -119,11 +162,45 @@ def plot_energy_against_n_vacuum_layer(
...
+    atom_height = get_n_th_slab_layer_height(atom, n_th_layer)
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:177]: final_height =

File: pseudopotential_calculator/calculations/slab.py [hunk 19]
```diff
@@ -119,11 +162,45 @@ def plot_energy_against_n_vacuum_layer(
...
+    height_per_slab_layer = get_height_per_slab_layer(atom)
+    atom_height_before_relax = (n_slab_layer - (n_th_layer - 1)) * height_per_slab_layer
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:179]: initial_height =

File: pseudopotential_calculator/calculations/slab.py [hunk 20]
```diff
@@ -119,11 +162,45 @@ def plot_energy_against_n_vacuum_layer(
...
         ("Vacuum Height", np.array(n_vacuum_layer), "N layers"),
         ("Energy", np.array(energies), "J"),
         ax=ax,
     )
+
+
+def get_atom_displacement(atom: Atoms, n_th_layer: int) -> float:
+    n_slab_layer = get_n_slab_layer(atom)
+    atom_height = get_n_th_slab_layer_height(atom, n_th_layer)
+    height_per_slab_layer = get_height_per_slab_layer(atom)
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:178]: height_per_layer is just as descriptive and shorter

File: pseudopotential_calculator/calculations/slab.py [hunk 21]
```diff
@@ -119,11 +162,45 @@ def plot_energy_against_n_vacuum_layer(
...
+    """Plot displacement of free atoms from it's initial position."""
+    displacement = list[float]()
+    n_free_layer = list[int]()
+    for calculator in calculators:
+        atom = get_calculator_atom(calculator)
+
+        displacement.append(
+            get_atom_displacement(atom, layer_idx),
+        )
+        # TODO fix potential bugs with data
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:198]: still need to fix this...

File: pseudopotential_calculator/calculations/slab.py [hunk 22]
```diff
@@ -119,11 +162,45 @@ def plot_energy_against_n_vacuum_layer(
...
+        n_free_layers = len(atom.constraints[1].index)  # type: ignore bad lin
+        n_free_layer.append(n_free_layers)
+    fig, ax, line = plot_data_comparison(
+        ("N Free Layers", np.array(n_free_layer), None),
+        ("Displacement", np.array(displacement), "m"),
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:203]: units

File: pseudopotential_calculator/calculations/slab.py [hunk 23]
```diff
@@ -78,32 +79,74 @@ def _get_height_per_layer(
     return np.abs(cast(float, position_0) - cast(float, position_1))
 
 
+def _get_constraints(
+    atoms: Atoms,
+    n_free_layer: int,
+) -> list[FixAtoms | FixedLine]:
+    heights = _get_heights_sorted(atoms)
+    sorted_indices = np.argsort(heights)
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:87]: no need to argsort here if heights are sorted?

File: pseudopotential_calculator/calculations/slab.py [hunk 24]
```diff
@@ -78,32 +79,74 @@ def _get_height_per_layer(
     return np.abs(cast(float, position_0) - cast(float, position_1))
 
 
+def _get_constraints(
+    atoms: Atoms,
+    n_free_layer: int,
+) -> list[FixAtoms | FixedLine]:
+    heights = _get_heights_sorted(atoms)
```
Comment (Matt-Ord) [inline pseudopotential_calculator/calculations/slab.py:86]: always define before use - python should be read top to bottom

**Review by Matt-Ord — CHANGES_REQUESTED** [review]

**Review by Matt-Ord — APPROVED** [review]

**Pull request merged**